File: java/proverb/src/main/java/Proverb.java
import java.util.stream.IntStream;

class Proverb {
    
  private final String[] words;

  Proverb(String[] words) {
    this.words = words;   
  }

  String recite() {
    return (words.length == 0) ? "" : 
      IntStream
        .range(0, words.length - 1)
        .mapToObj(i -> "For want of a " + words[i] + " the " + words[i + 1] + " was lost.")
        .reduce("", (a,b) -> a + b + "\n")
        .concat("And all for the want of a " + words[0] + ".");
  }
}

File: csharp/proverb/Proverb.cs
using System.Linq;

public static class Proverb
{
  public static string[] Recite(string[] words) => 
    !words.Any() ? new string[0] :
    Enumerable.Concat
    (
       Enumerable
        .Range(0, words.Length - 1)
        .Select(i => $"For want of a {words[i]} the {words[i + 1]} was lost."),

       new[] { $"And all for the want of a {words[0]}." }
    )
    .ToArray();
}